README.md

```diff
@@ -247,6 +247,49 @@ concat(42, 13)
 // at position 1 does not.
 ```
 
+<a name="type-constructors"></a>
+### Type constructors
+_Added in v0.3.0_
+
+If you need type constructors, that could have more than a fixed number of combinations
+of Types, you should not add it into environments, cause it will produce errors.
+You should add it into special field `typeConstructors`.
+
+Type constructors is a function, that will return Type. And it will be used by a `hm-def`
+to define exactly needed Type in the function, without adding it into global environments.
+
+```javascript
+const Either = $.BinaryType(
+  'my-package/Either',
+  'http://example.com/my-package#Either',
+  x => x != null && x['@@type'] === 'my-package/Either'
+  either => (either.isLeft ? [either.value] : []),
+  either => (either.isRight ? [either.value] : [])
+);
+// Either is a function `Either :: Type -> Type -> Type`,
+
+const def = HMD.create({
+  checkTypes: true,
+  env: $.env,
+  typeConstructors: [
+    Either,
+  ],
+});
```

Comment:
Why `typeConstructors: [Either]` here?

Why not simply <code>env: $.env.concat([Either([$.Unknown][1], [$.Unknown][1])])</code>?


[1]: https://github.com/sanctuary-js/sanctuary-def#Unknown

Reply:
There are a couple of problems:
1. In this case, we'll get Type `Either a b`, but when I define and call function `foo :: Either Number String -> Either String String`, it produces an Error, cause it returns not the same type `a`.
2. In addition, the error message becomes difficult to read and not quite true.
![2018-01-10 15 55 04](https://user-images.githubusercontent.com/1897530/34774169-6536f4ac-f61f-11e7-9855-54bec9cfdb58.png)

If you have ideas how to solve these problems — happy to hear.

Reply:
You can include `Either($.Unknown, $.Unknown)` in the environment to appease sanctuary-def, but you'll need to replace those placeholder types with actual types when using `Either :: Type -> Type -> Type` with `def`. You should be able to do this be accessing properties of `Either($.Unknown, $.Unknown)` and using these values as arguments to [`$.BinaryType`][1]. See [`xprod`][2] for inspiration. It's doing this specialization for every combination of types. In your case you may only need to perform one specialization. `t` in your case could be `Either($.Unknown, $.Unknown)`. Does this make sense?


[1]: https://github.com/sanctuary-js/sanctuary-def#BinaryType
[2]: https://github.com/sanctuary-js/sanctuary-def/blob/v0.14.0/index.js#L1609-L1623

Reply:
Don't quite understand what you mean.
In the example above, I added `Either($.Unknown, $.Unknown)` in `env`. And in the function definition it creates a specific type `Either($.Number $.String)` and `Either($.String $.String)`. However, we get the result that I posted above.
So, what I'm doing wrong?

Reply:
I will investigate, but I have a lot going on today so I won't have an opportunity until tomorrow.

Reply:
Okay, thanks.
And maybe you could answer, do we really need a Unary/BinaryType with Unknowns in `env` if we define specific Type in the function definition?

Reply:
> And maybe you could answer, do we really need a Unary/BinaryType with Unknowns in `env` if we define specific Type in the function definition?

The environment is required only for polymorphic functions such as `id :: a -> a`. If you wish to support such functions, you will need to ensure that the environment contains any types you wish to support. Otherwise `id(Right(42))` would be an error because `Right(42)` would not be a member of any type in the environment.

Reply:
Thanks! I got it.
Got refactored it and get rid of `typeConstructors` property (recreate specific Unary/Binary Types from the same with `$.Unknown` types).
I will be glad if you rate the correctness of the solution.